include/xtensor/xhistogram.hpp

```diff
@@ -0,0 +1,307 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Sylvain Corlay and Wolf Vollprecht    *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+/**
+ * @brief construct histogram
+ */
+
+#ifndef XTENSOR_HISTOGRAM_HPP
+#define XTENSOR_HISTOGRAM_HPP
+
+#include "xtensor.hpp"
+#include "xsort.hpp"
+
+namespace xt
+{
+
+    /**
+     * @ingroup basic_functions
+     * @brief Compute the histogram of a set of data.
+     *
+     * @param The data.
+     * @param The bin-edges.
+     * @param Weight factors corresponding to each data-point.
+     * @param If true the resulting integral is normalized to 1. [default: false]
+     * @return An one-dimensional xarray<double>, length: bin_edges.size()-1.
+     */
+    template<class E1, class E2, class E3>
+    inline auto histogram(
+        const E1 &data, const E2 &bin_edges, const E3 &weights, bool density=false
```

Comment:
Universal references would allow to directly move temporaries instead of saving them in an intermediate variable

Reply:
@JohanMabille I'm having trouble getting the syntax right, in particular extracting a typename. I tried to find inspiration elsewhere in the code, but that did not quite bring it home:

Reply:
From what I see in the test, `data` is a container of values right? So I don't understand why you need an intermediate `accumulator_functor` nor what it is. I think the following should be enough:
```cpp
using value_type = typename std::decay_t<E1>::value_type;
```

Reply:
Universal reference for E1, E2 and E3 still missing